### PR TITLE
SciPyPrinter: polygamma (fix gh-23924)

### LINKED ISSUE #23924: CI failure: lambdify: NameError: name 'polygamma' is not defined
This failure is seen in CI for the 3.11 optional dependency tests:
```
____________ sympy/utilities/tests/test_lambdify.py:test_scipy_fns _____________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/utilities/tests/test_lambdify.py", line 1103, in test_scipy_fns
    assert abs(f(tv) - sympy_result) < 1e-13*(1 + abs(sympy_result))
               ^^^^^
  File "<lambdifygenerated-604>", line 2, in _lambdifygenerated
    return polygamma(0, x)
           ^^^^^^^^^
NameError: name 'polygamma' is not defined
```
That CI job uses latest cython, numpy and scipy from git so this is likely due to a change in scipy.

### PR DESCRIPTION
**Pull request by bjodah**
#### References to other Issues or PRs
~~Fixes #23924~~

#### Brief description of what is fixed or changed
~~Attempt to fix CI failure.~~
Add polygamma to explicit list of functions that SciPy can evaluate.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/utilities/tests/test_lambdify.py [hunk 1]
```diff
@@ -1077,7 +1077,7 @@ def test_scipy_fns():
     if not scipy:
         skip("scipy not installed")
 
-    single_arg_sympy_fns = [Ei, erf, erfc, factorial, gamma, loggamma, digamma]
+    single_arg_sympy_fns = [Ei, erf, erfc, factorial, gamma, loggamma, digamma, polygamma]
```
Comment (oscargus) [inline sympy/utilities/tests/test_lambdify.py:1080]: I have no idea how this test works, but does this "guarantee" that polygamma works when both arguments are used? (Just considering the variable name...)

Comment (bjodah) [reply sympy/utilities/tests/test_lambdify.py]: No, I didn't read the test code carefully enough, next commit now passes for me locally.

File: sympy/printing/numpy.py [hunk 2]
```diff
@@ -290,6 +290,7 @@ def _print_NDimArray(self, expr):
     'gamma': 'gamma',
     'loggamma': 'gammaln',
     'digamma': 'psi',
+    'polygamma': 'polygamma',
```
Comment (oscarbenjamin) [inline sympy/printing/numpy.py:293]: Why is this necessary now if it wasn't necessary before? Has something changed in scipy?

Comment (bjodah) [reply sympy/printing/numpy.py:293]: I don't actually know. `lambdify` uses `exec` with a namespace which I think contains `__all__` from the imported module. So I guess either the logic in `lambdify` changed or the exported functions in SciPy changed?

Comment (bjodah) [reply sympy/printing/numpy.py:293]: On master it looks like this:
```
Created with lambdify. Signature:                                                                                                                                      
                                                                                                                                                                       
func(x, y)                                                                                                                                                             
                                                                                                                                                                       
Expression:                                                                                                                                                            
                                                                                                                                                                       
polygamma(x, y)                                                                    
                                                                                                                                                                       
Source code:                      
                                                                                   
def _lambdifygenerated(x, y):                                                      
    return polygamma(x, y)                                                         
                                                                                   
                                                                                   
Imported modules:       

```

and I expected the function call to be qualified when using this branch, but actually it is not... So now I'm just confused.

Comment (oscarbenjamin) [reply sympy/printing/numpy.py:293]: I've been struggling to build scipy myself to test this out. Does `from sympy import *` import polygamma with scipy main? Is that what changed since the 1.9.1 release?

Comment (oscarbenjamin) [reply sympy/printing/numpy.py:293]: Part of the idea of testing with upstream git branches is not just so that SymPy can detect changes and respond accordingly but also to detect regressions in the upstream codebases that might impact SymPy. In other words before changing anything in SymPy we should establish whether this looks like an intentional change in SciPy and if it should be considered a bug in SciPy then it should be reported there so it can be fixed before their next release.

Comment (bjodah) [reply sympy/printing/numpy.py:293]: Makes sense. I can't seem to reproduce this using Python 3.11 and SciPy-git built a few weeks back:
```
Python 3.11.0b3 (main, Jun 16 2022, 13:59:02) [GCC 12.0.1 20220319 (experimental) [master r12-7719-g8ca61ad148f]]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import scipy

In [2]: from sympy import *

In [3]: from sympy.abc import x,y

In [4]: f = lambdify([x,y], polygamma(x,y))

In [5]: print(f.__doc__)
Created with lambdify. Signature:

func(x, y)

Expression:

polygamma(x, y)

Source code:

def _lambdifygenerated(x, y):
    return polygamma(x, y)


Imported modules:



In [6]: f(1,0.5)
Out[6]: array(4.9348022)

In [7]: scipy.__version__
Out[7]: '1.10.0.dev0+1303.1ef52c6'
```

Comment (oscarbenjamin) [reply sympy/printing/numpy.py:293]: > SciPy-git built a few weeks back:

Looking back through CI the test started failing 3 days ago so we need a more recent SciPy.

Comment (bjodah) [reply sympy/printing/numpy.py:293]: Thanks, I'll recompile current SciPy master and see what that gets us.

Comment (oscarbenjamin) [reply sympy/printing/numpy.py:293]: Thanks. It could also be a change in numpy. The setup can be seen here:
https://github.com/sympy/sympy/blob/b1f35f8daf5efa9f801f905cf1037000d69b0278/.github/workflows/runtests.yml#L186-L215

Comment (bjodah) [reply sympy/printing/numpy.py:293]: Alright, so using SciPy master I can reproduce the failure:
```
Python 3.11.0b3 (main, Jun 16 2022, 13:59:02) [GCC 12.0.1 20220319 (experimental) [master r12-7719-g8ca61ad148f]]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sympy

In [2]: import scipy

In [3]: from sympy.abc import x,y

In [5]: f = sympy.lambdify([x,y], sympy.polygamma(x,y))

In [6]: print(f.__doc__)
Created with lambdify. Signature:

func(x, y)

Expression:

polygamma(x, y)

Source code:

def _lambdifygenerated(x, y):
    return polygamma(x, y)


Imported modules:



In [7]: f(1,0.5)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [7], in <cell line: 1>()
----> 1 f(1,0.5)

File <lambdifygenerated-1>:2, in _lambdifygenerated(x, y)
      1 def _lambdifygenerated(x, y):
----> 2     return polygamma(x, y)

NameError: name 'polygamma' is not defined

In [8]: scipy
Out[8]: <module 'scipy' from '/opt/cpython-3.11.0b3-release/lib/python3.11/site-packages/scipy/__init__.py'>

In [9]: scipy.__version__, sympy.__version__
Out[9]: ('1.10.0.dev0+1668.ac729b8', '1.10.1')
```

Comment (bjodah) [reply sympy/printing/numpy.py:293]: And using my branch in this PR does not solve this issue.

Comment (oscarbenjamin) [reply sympy/printing/numpy.py:293]: Can you bisect to find the relevant scipy commit?

Comment (bjodah) [reply sympy/printing/numpy.py:293]: Manually bisecting SciPy commits:
b5e854576b39c784b2c868bd6835b378845511bf OK
d58e74655e1129d5d18f034b5a0219dba90368d3 bad

Looking in scipy/scipy#15607 I see L68-L82 in [`scipy/datasets/__init__.py`](https://github.com/scipy/scipy/pull/15607/files#diff-89f9e47b57fc95d026c7ba70e5915600d02abb9b62a26b2b381d817ac0581160) which looks suspicious.

And indeed: running `pip install pooch` fixes our issue.

**Comment by oscarbenjamin** [comment]
Looks like this works. I fixed the slow2 test failure in #23944

Is it still necessary to add polygamma explicitly or can that part of the PR be reverted now?

**Comment by oscarbenjamin** [comment]
I see there are several changes here. Maybe it would be better to have one PR that just removes `from scipy import *`. I'll open that as a separate PR.

**Comment by bjodah** [comment]
OK, so at this point this PR extends our testing lambdify with of `polygamma` and makes sure that the `SciPyPrinter` knows that the function exists in `scipy.special`.

**Comment by bjodah** [comment]
> Is it still necessary to add polygamma explicitly or can that part of the PR be reverted now?
it should be added explicitly, if nothing else for consistency. (it doesn't matter in lambdify though since it actively disables fully qualified function names)
